web_poet/rules.py

```diff
@@ -0,0 +1,393 @@
+from __future__ import annotations  # https://www.python.org/dev/peps/pep-0563/
+
+import importlib
+import importlib.util
+import pkgutil
+import warnings
+from collections import deque
+from operator import attrgetter
+from typing import Any, Dict, Iterable, List, Optional, Type, TypeVar, Union
+
+import attrs
+from url_matcher import Patterns
+
+from web_poet._typing import get_item_cls
+from web_poet.pages import ItemPage
+from web_poet.utils import _create_deprecated_class, as_list, str_to_pattern
+
+Strings = Union[str, Iterable[str]]
+
+PageObjectRegistryTV = TypeVar("PageObjectRegistryTV", bound="PageObjectRegistry")
+
+
+@attrs.define(frozen=True)
+class ApplyRule:
+    """A rule that primarily applies Page Object and Item overrides for a given
+    URL pattern.
+
+    This is instantiated when using the :func:`web_poet.handle_urls` decorator.
+    It's also being returned as a ``List[ApplyRule]`` when calling the
+    ``web_poet.default_registry``'s :meth:`~.PageObjectRegistry.get_rules`
+    method.
+
+    You can access any of its attributes:
+
+        * ``for_patterns`` - contains the list of URL patterns associated with
+          this rule. You can read the API documentation of the `url-matcher
+          <https://url-matcher.readthedocs.io/>`_ package for more information
+          about the patterns.
+        * ``use`` - The Page Object that will be **used** in cases where the URL
+          pattern represented by the ``for_patterns`` attribute is matched.
+        * ``instead_of`` - *(optional)* The Page Object that will be **replaced**
+          with the Page Object specified via the ``use`` parameter.
+        * ``to_return`` - *(optional)* The item class that marks the Page Object
+          to be **used** which is capable of returning that item class.
+        * ``meta`` - *(optional)* Any other information you may want to store.
+          This doesn't do anything for now but may be useful for future API updates.
+
+    The main functionality of this class lies in the ``instead_of`` and ``to_return``
+    parameters. Should both of these be omitted, then :class:`~.ApplyRule` simply
+    tags which URL patterns the given Page Object defined in ``use`` is expected
+    to be used. It works as:
+
+        1. Given a URL, match it against the ``for_patterns`` from the registry
+           rules.
+        2. This could give us a collection of rules. We need to select one based
+           on the highest priority set by `url-matcher`_.
+        3. When a single rule has been selected, use the the Page Object specified
+           in its ``use`` parameter.
+
+    If ``instead_of=None``, this simply means that the Page Object assigned in
+    the ``use`` parameter will be utilized for all URLs matching the URL pattern
+    in ``for_patterns``. However, if ``instead_of=ReplacedPageObject``, then it
```

Comment:
This doesn't look correct: if `instead_of=None`, then Page Object won't be used to replace another Page Object, regardless of URL pattern. It might be used on the provided URL pattern if `to_return` is set, but this has nothing to do with `instead_of=None`, as it should work regardless of `instead_of` value. 

So, in the example below:

```py
@handle_urls("example.com")
class MyPage(ItemPage):
    # ...
``` 

handle_urls is a pure documentation. You can't use it instead of any other page object (as instead_of is not provided), and you can't use it to return a certain data type (because the returned item class is unknown).

Reply:
@BurnzZ let me send a PR with some doc-related suggestions

Reply:
https://github.com/scrapinghub/web-poet/pull/90